src/net_processing.cpp

```diff
@@ -2443,15 +2443,9 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, const Peer& peer,
         return;
     }
 
-    {
-        LOCK(cs_main);
-
-        // If we don't have the last header, then they'll have given us
-        // something new (if these headers are valid).
-        if (!m_chainman.m_blockman.LookupBlockIndex(headers.back().GetHash())) {
-            received_new_header = true;
-        }
-    }
+    // If we don't have the last header, then this peer will have given us
+    // something new (if these headers are valid).
+    received_new_header = WITH_LOCK(::cs_main, return m_chainman.m_blockman.LookupBlockIndex(headers.back().GetHash()) == nullptr);
```

Comment:
We need to do the calculation here before invoking `ProcessNewBlockHeaders()` because that will cause headers to be added to the block index, but I'll get rid of the declaration at line 2423, as we can just declare and initialize here instead.